# *Analysed By Eng.Mohamed Abdullah Ali*

# Phase 1 — Problem Definition & KPI Setup

## 1.1 Business Context

DataCo operates across multiple areas of the supply chain, including sales, customers, products, shipping, and order fulfillment.

The objective of this project is to evaluate the company's commercial and operational performance and identify the major factors that may be affecting:

- Delivery reliability
- Transaction risk
- Profitability
- Customer experience
- Overall supply-chain performance

The analysis will combine business understanding, data quality assessment, exploratory analysis, SQL analysis, data modeling, and Power BI reporting to transform raw data into actionable business insights.

> **Important:** The problems identified in this phase are initial business hypotheses. They will be validated against the dataset during the Data Understanding and Data Quality phases before any KPI is considered reliable.

## 1.2 Business Problem

The company is potentially facing three major business challenges:

### 1. Late Delivery Risk

Orders may not be delivered according to the planned or scheduled delivery timeline.

Late deliveries can negatively affect:

- Customer satisfaction
- Customer experience
- Operational efficiency
- Brand perception
- Future customer behavior

The analysis should determine the magnitude of the problem and identify the markets, products, shipping methods, and other factors associated with delivery delays.

---


### 2. Profitability Erosion

The company may be losing potential profitability because of factors such as:

- High discounts
- Shipping costs
- Delivery problems
- Returns or other operational costs
- Low-margin products or categories

The objective is not to assume that these factors cause profit erosion, but to investigate whether they are associated with lower profitability and identify where profitability is being lost.

## 1.3 Business Objective

The main objective of this project is to develop a data-driven view of DataCo's commercial and operational performance and provide management with actionable insights.

### Specific Objectives

1. Measure overall sales and profitability performance.
2. Quantify the level of late-delivery risk.
3. Assess fraudulent or suspected-risk transactions based on the available data.
4. Identify products, categories, markets, and customer segments with strong or weak profitability.
5. Investigate the relationship between discounts and profitability.
6. Evaluate shipping and delivery performance.
7. Identify operational areas associated with higher delivery risk.
8. Investigate factors associated with customer experience and satisfaction.
9. Identify the major sources of commercial and operational risk.
10. Translate analytical findings into practical business recommendations.

### Final Business Goal

> **Identify where the company is losing revenue, profitability, operational efficiency, or customer value and determine where management should prioritize corrective action.**

## 1.4 Initial Business Hypotheses

Before analyzing the data, we establish the following hypotheses.

### Hypothesis 1 — Delivery Performance

> A significant proportion of orders may experience late delivery, with the risk potentially concentrated in specific markets, shipping modes, products, or operational segments.

### Hypothesis 2 — Transaction Risk

> Fraudulent or high-risk transactions may be concentrated in specific customer, product, geographic, or payment segments.

### Hypothesis 3 — Discount & Profitability

> Higher discount levels may be associated with lower profit margins, particularly for specific products or categories.

### Hypothesis 4 — Shipping & Profitability

> Higher shipping costs and operational inefficiencies may contribute to lower order profitability.

### Hypothesis 5 — Delivery & Customer Experience

> Poor delivery performance may be associated with lower customer satisfaction and review scores.

> These hypotheses are not conclusions. They will be tested and either supported, rejected, or found inconclusive based on the available evidence.

# 1.5 KPI Framework

The KPI framework is divided into three major business areas:

| Business Area | Primary KPI | Supporting KPIs | Main Analytical Dimensions |
|---|---|---|---|
| **Late Delivery Risk** | Late Delivery Rate | Late Orders, On-Time Rate, Shipping Delay | Market, Region, Shipping Mode, Category, Seller, Time |
| **Fraud / Transaction Risk** | Risk/Fraud Rate | Risky Orders, Exposed Revenue, Exposed Profit | Market, Payment Type, Customer, Product, Time |
| **Profitability Erosion** | Profit Margin | Total Profit, Profit/Order, Discount, Shipping Cost | Product, Category, Market, Customer, Shipping Mode |

> **KPI definitions will be finalized after validating the dataset's grain, fields, relationships, and business definitions.**

## 1.6 KPI Definitions

### A. Late Delivery Rate

Measures the percentage of eligible orders that are classified as late.

**Conceptual definition:**

Late Delivery Rate = Late Orders / Total Eligible(Arri) Orders × 100

**Business question:**

> How significant is the company's delivery-risk problem?

**Diagnostic dimensions:**

- Time
- Market
- Region
- Shipping Mode
- Product Category
- Seller
- Customer Segment

---

### B. Risk / Fraud Rate

Measures the proportion of transactions identified as fraudulent or risky according to the validated dataset definition.

**Conceptual definition:**

Risk/Fraud Rate = Risky or Fraudulent Orders / Total Eligible Orders × 100

**Business question:**

> How large is the company's transaction-risk exposure?

**Diagnostic dimensions:**

- Market
- Payment Type
- Customer
- Product
- Time
- Order characteristics

> The exact definition will depend on the validated meaning of the fraud/risk field.

---

### C. Total Profit

Measures the total profit generated during the analyzed period.

**Conceptual definition:**

Total Profit = Sum of Validated Order Profit

**Business question:**

> How much profit is the business generating?

---

### D. Profit Margin

Measures profitability relative to sales.

**Conceptual definition:**

Profit Margin = Total Profit / Total Sales × 100

**Business question:**

> Is revenue growth translating into profitable growth?

---

### E. Average Profit per Order

Measures the average profit generated by each eligible order.

**Conceptual definition:**

Average Profit per Order = Total Profit / Total Eligible Orders

**Business question:**

> How much profit does the average order generate?

---

### F. Discount Impact

The analysis will investigate whether increasing discount levels are associated with declining profitability.

**Business question:**

> At what point does discounting begin to significantly reduce profitability?

---

### G. Shipping Cost Impact

The analysis will investigate whether shipping costs are associated with lower order profitability.

**Business question:**

> Which shipping-related factors are contributing to profitability erosion?

# 1.7 KPI Hierarchy

To avoid creating a dashboard full of disconnected numbers, KPIs will be organized into three analytical levels.

### Level 1 — Executive KPIs

These provide a high-level view of business performance.

- Total Sales
- Total Profit
- Profit Margin
- Total Orders
- Late Delivery Rate
- Risk/Fraud Rate

### Level 2 — Diagnostic KPIs

These help explain changes in the executive KPIs.

- Late Orders
- On-Time Delivery Rate
- Average Shipping Delay
- Average Profit per Order
- Discount Rate
- Shipping Cost
- Risk/Fraud Exposure

### Level 3 — Root-Cause Dimensions

These allow management to investigate where problems are concentrated.

- Time
- Market
- Region
- Country
- Product
- Product Category
- Customer Segment
- Shipping Mode
- Payment Type
- Seller
- Order Priority

### Analytical Flow

Executive KPI
→ Problem Area
→ Diagnostic KPI
→ Root-Cause Dimension
→ Business Insight
→ Business Recommendation

# 1.8 Key Management Questions

The project will ultimately answer the following questions:

### Commercial Performance

1. Is the company growing its sales and profit over time?
2. Which products and categories generate the highest profit?
3. Which products generate high sales but weak profitability?
4. Which markets contribute most to profitable growth?

### Delivery & Operations

5. What percentage of orders are delivered late?
6. Which markets have the highest delivery risk?
7. Which shipping methods are associated with higher delivery delays?
8. Which products or categories experience the greatest delivery problems?

### Transaction Risk

9. What proportion of transactions are classified as risky or fraudulent?
10. Where is transaction risk concentrated?
11. What is the financial exposure associated with risky transactions?

### Profitability

12. How does discounting affect profitability?
13. Which products or categories have weak profit margins?
14. How do shipping costs affect order profitability?
15. Where is the greatest profitability erosion occurring?

### Customer Experience

16. Is delivery performance associated with customer satisfaction?
17. Which operational factors are associated with lower customer ratings?

### Strategic Decision-Making

18. Which problems should management prioritize?
19. Which products, markets, or operational areas represent the greatest opportunity?
20. What actions could improve profitability while reducing operational risk?

# 1.9 Analytical Decision Framework

The final objective is not simply to report KPIs.

We will follow the analytical chain:

**Business Problem**
↓
**Business Question**
↓
**Required KPI**
↓
**Data Validation**
↓
**Analysis**
↓
**Finding**
↓
**Business Insight**
↓
**Recommendation**
↓
**Management Decision**

For example:

> High Late Delivery Rate  
> → Identify affected markets  
> → Identify affected shipping modes  
> → Investigate operational drivers  
> → Quantify business impact  
> → Recommend corrective action

This ensures that every analytical result has a clear business purpose.

# 1.10 Phase 1 Conclusion

At this stage, three potential business problems have been identified:

1. **Late Delivery Risk**
2. **Fraud / Transaction Risk**
3. **Profitability Erosion**

These represent our initial analytical priorities.

However, no KPI or business conclusion will be considered final until the dataset has been validated.

### Next Phase

**Phase 2 — Data Understanding**

We will investigate:

- Dataset structure
- Number of records and features
- Data types
- Data dictionary
- Missing values
- Duplicate records
- Unique identifiers
- Data grain
- Primary-key candidates
- Foreign-key candidates
- Relationships between entities
- Definitions of revenue, profit, shipping, delivery, and risk fields
- Whether each proposed KPI can actually be calculated reliably

> **Principle:** We will never manufacture a business metric from an assumption that the data cannot support.

# Phase 2 — Data Understanding

## 2.1 Dataset Overview

## 2.2 Dataset Structure

## 2.3 Data Dimensions

## 2.4 Data Types

## 2.5 Data Dictionary & Business Meaning

## 2.6 Entity Identification

## 2.7 Data Grain

## 2.8 Identifier & Uniqueness Assessment

## 2.9 Missing Values Overview

## 2.10 Duplicate Records

## 2.11 Cardinality & Relationship Exploration

## 2.12 Numerical Feature Overview

## 2.13 Categorical Feature Overview

## 2.14 Temporal Coverage

## 2.15 Initial KPI Feasibility

## 2.16 Phase 2 Summary
<hr></hr>

### *2.1 Dataset Overview*

- First, we load the actual DataCo file you're working with.

In [5]:
import pandas as pd
df = pd.read_csv(
    "DataCoSupplyChainDataset.csv",
    encoding="latin1"
)
pd.set_option('display.max_columns', None) # 'None' tells Pandas to show all columns

In [6]:
df.shape

(180519, 53)

In [7]:
df.head()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,725.0,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,1/31/2018 22:56,77202,1360,13.110000,0.04,180517,327.75,0.29,1,327.75,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,2/3/2018 22:56,Standard Class
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,725.0,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,1/13/2018 12:27,75939,1360,16.389999,0.05,179254,327.75,-0.80,1,327.75,311.359985,-249.089996,South Asia,Rajastán,PENDING,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/18/2018 12:27,Standard Class
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,95125.0,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,1/13/2018 12:06,75938,1360,18.030001,0.06,179253,327.75,-0.80,1,327.75,309.720001,-247.779999,South Asia,Rajastán,CLOSED,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/17/2018 12:06,Standard Class
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,EE. UU.,XXXXXXXXX,Tana,19490,Tate,XXXXXXXXX,Home Office,CA,3200 Amber Bend,90027.0,2,Fitness,34.125946,-118.291016,Pacific Asia,Townsville,Australia,19490,1/13/2018 11:45,75937,1360,22.940001,0.07,179252,327.75,0.08,1,327.75,304.809998,22.860001,Oceania,Queensland,COMPLETE,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/16/2018 11:45,Standard Class
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Orli,19489,Hendricks,XXXXXXXXX,Corporate,PR,8671 Iron Anchor Corners,725.0,2,Fitness,18.253769,-66.037048,Pacific Asia,Townsville,Australia,19489,1/13/2018 11:24,75936,1360,29.500000,0.09,179251,327.75,0.45,1,327.75,298.250000,134.210007,Oceania,Queensland,PENDING_PAYMENT,NaN,1360,73,NaN,http://images.acmesports.sports/Smart+watch,Smart watch,327.75,0,1/15/2018 11:24,Standard Class


In [8]:
df.tail()

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode
180514,CASH,4,4,40.000000,399.980011,Shipping on time,0,45,Fishing,Brooklyn,EE. UU.,XXXXXXXXX,Maria,1005,Peterson,XXXXXXXXX,Home Office,NY,1322 Broad Glade,11207.0,7,Fan Shop,40.640930,-73.942711,Pacific Asia,Shanghái,China,1005,1/16/2016 3:40,26043,1004,0.0,0.00,65177,399.980011,0.10,1,399.980011,399.980011,40.000000,Eastern Asia,Shanghái,CLOSED,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,1/20/2016 3:40,Standard Class
180515,DEBIT,3,2,-613.770019,395.980011,Late delivery,1,45,Fishing,Bakersfield,EE. UU.,XXXXXXXXX,Ronald,9141,Clark,XXXXXXXXX,Corporate,CA,7330 Broad Apple Moor,93304.0,7,Fan Shop,35.362545,-119.018700,Pacific Asia,Hirakata,Japón,9141,1/16/2016 1:34,26037,1004,4.0,0.01,65161,399.980011,-1.55,1,399.980011,395.980011,-613.770019,Eastern Asia,Osaka,COMPLETE,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,1/19/2016 1:34,Second Class
180516,TRANSFER,5,4,141.110001,391.980011,Late delivery,1,45,Fishing,Bristol,EE. UU.,XXXXXXXXX,John,291,Smith,XXXXXXXXX,Corporate,CT,97 Burning Landing,6010.0,7,Fan Shop,41.629959,-72.967155,Pacific Asia,Adelaide,Australia,291,1/15/2016 21:00,26024,1004,8.0,0.02,65129,399.980011,0.36,1,399.980011,391.980011,141.110001,Oceania,Australia del Sur,PENDING,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,1/20/2016 21:00,Standard Class
180517,PAYMENT,3,4,186.229996,387.980011,Advance shipping,0,45,Fishing,Caguas,Puerto Rico,XXXXXXXXX,Mary,2813,Smith,XXXXXXXXX,Consumer,PR,2585 Silent Autumn Landing,725.0,7,Fan Shop,18.213350,-66.370575,Pacific Asia,Adelaide,Australia,2813,1/15/2016 20:18,26022,1004,12.0,0.03,65126,399.980011,0.48,1,399.980011,387.980011,186.229996,Oceania,Australia del Sur,PENDING_PAYMENT,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,1/18/2016 20:18,Standard Class
180518,PAYMENT,4,4,168.949997,383.980011,Shipping on time,0,45,Fishing,Caguas,Puerto Rico,XXXXXXXXX,Andrea,7547,Ortega,XXXXXXXXX,Consumer,PR,697 Little Meadow,725.0,7,Fan Shop,18.290380,-66.370613,Pacific Asia,Nagercoil,India,7547,1/15/2016 18:54,26018,1004,16.0,0.04,65113,399.980011,0.44,1,399.980011,383.980011,168.949997,South Asia,Tamil Nadu,PENDING_PAYMENT,NaN,1004,45,NaN,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,1/19/2016 18:54,Standard Class


### *2.2 Dataset Structure*

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180519 entries, 0 to 180518
Data columns (total 53 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   Type                           180519 non-null  object 
 1   Days for shipping (real)       180519 non-null  int64  
 2   Days for shipment (scheduled)  180519 non-null  int64  
 3   Benefit per order              180519 non-null  float64
 4   Sales per customer             180519 non-null  float64
 5   Delivery Status                180519 non-null  object 
 6   Late_delivery_risk             180519 non-null  int64  
 7   Category Id                    180519 non-null  int64  
 8   Category Name                  180519 non-null  object 
 9   Customer City                  180519 non-null  object 
 10  Customer Country               180519 non-null  object 
 11  Customer Email                 180519 non-null  object 
 12  Customer Fname                

### *2.3 Data Dimensions*

In [10]:
rows, columns = df.shape

print(f"Rows: {rows:,}")
print(f"Columns: {columns:,}")

Rows: 180,519
Columns: 53


### Dataset Dimensions

The dataset contains **180,519 records and 53 features**.

This represents the initial analytical dataset before any transformation or filtering.

### *2.4 Data Types*

In [11]:
df.dtypes

Type                              object
Days for shipping (real)           int64
Days for shipment (scheduled)      int64
Benefit per order                float64
Sales per customer               float64
Delivery Status                   object
Late_delivery_risk                 int64
Category Id                        int64
Category Name                     object
Customer City                     object
Customer Country                  object
Customer Email                    object
Customer Fname                    object
Customer Id                        int64
Customer Lname                    object
Customer Password                 object
Customer Segment                  object
Customer State                    object
Customer Street                   object
Customer Zipcode                 float64
Department Id                      int64
Department Name                   object
Latitude                         float64
Longitude                        float64
Market          

In [12]:
dtype_summary = (
    df.dtypes
      .value_counts()
      .rename_axis("Data Type")
      .reset_index(name="Number of Columns")
)

dtype_summary

,Data Type,Number of Columns
0,object,24
1,float64,15
2,int64,14


### *2.5 Data Dictionary & Business Meaning*

*Now we map the columns into business groups.*
<br>
Based on the DataCo dataset structure, we expect categories such as:

### Order
- Order Id
- Order Date
- Order Status
- Order Region
- Order Country
- Order City
- Order State
- Order Item Quantity
### Customer
- Customer Id
- Customer Name
- Customer Segment
- Customer Country
- Customer City
### Product
- Product Card Id
- Product Name
- Product Category Id
- Category Name
- Product Price
### Shipping
- Shipping Mode
- Days for Shipping (Real)
- Days for Shipment (Scheduled)
- Delivery Status
- Late Delivery Risk
### Financial
- Sales
- Order Item Total
- Order Item Profit Ratio
- Benefit per Order
- Order Profit Per Order
### Risk
- Order Status
- Late Delivery Risk
- Potential fraud-related fields

### *2.6 Entity Identification*

*What real-world entities exist in this dataset?* <br>
- Customer
- Product
- Order
- Order Item
- Seller
- Shipping
- Geography

### *2.7 Data Grain*

*What does one row represent?*
<br>

The identifier analysis indicates that the dataset contains **180,519 unique Order Item IDs across 180,519 records**, meaning that `Order Item Id` uniquely identifies each transactional row.

At the same time, the dataset contains only **65,752 distinct Order IDs**, demonstrating that a single order can contain multiple order-item records.

The dataset also contains:

- **20,652 distinct customers**
- **65,752 distinct orders**
- **180,519 order items**
- **118 distinct products**

Therefore, the initial analytical grain of the dataset is:

> **One row represents one Order Item / Order Line.**

This distinction is critical for subsequent analysis because row count cannot be used directly as the number of orders.

For example:

- `COUNT(Order Item Id)` → Order-item volume
- `COUNT(DISTINCT Order Id)` → Order volume
- `COUNT(DISTINCT Customer Id)` → Customer volume
- `COUNT(DISTINCT Product Card Id)` → Product volume

This grain definition will also be essential when designing the analytical data model and calculating KPIs in Power BI.

### *2.8 Identifier & Uniqueness Assessment*

In [13]:
id_columns = [
    "Order Id",
    "Customer Id",
    "Product Card Id",
    "Order Item Id"
]

for col in id_columns:
    print(
        f"{col}: "
        f"{df[col].nunique():,} unique values / "
        f"{len(df):,} rows"
    )

Order Id: 65,752 unique values / 180,519 rows
Customer Id: 20,652 unique values / 180,519 rows
Product Card Id: 118 unique values / 180,519 rows
Order Item Id: 180,519 unique values / 180,519 rows


In [14]:
df["Order Item Id"].duplicated().sum()

np.int64(0)

In [15]:
order_item_counts = (
    df.groupby("Order Id")["Order Item Id"]
      .nunique()
)

order_item_counts.describe()

count    65752.000000
mean         2.745453
std          1.478363
min          1.000000
25%          1.000000
50%          3.000000
75%          4.000000
max          5.000000
Name: Order Item Id, dtype: float64

In [16]:
order_item_counts.value_counts().sort_index()

Order Item Id
1    19850
2    11447
3    11398
4    11704
5    11353
Name: count, dtype: int64

In [17]:
order_product_counts = (
    df.groupby("Order Id")["Product Card Id"]
      .nunique()
)

order_product_counts.describe()

count    65752.000000
mean         2.429782
std          1.257508
min          1.000000
25%          1.000000
50%          2.000000
75%          3.000000
max          5.000000
Name: Product Card Id, dtype: float64

In [18]:
order_product_counts.value_counts().sort_index()

Product Card Id
1    21174
2    14174
3    15102
4    11575
5     3727
Name: count, dtype: int64

In [19]:
df.groupby("Order Item Id")["Order Id"].nunique().max()

1

In [20]:
df.groupby("Order Item Id")["Product Card Id"].nunique().max()

1

In [21]:
same_product_multiple_items = (
    df.groupby(["Order Id", "Product Card Id"])["Order Item Id"]
      .nunique()
)

same_product_multiple_items.value_counts().sort_index()

Order Item Id
1    140733
2     17386
3      1563
4        80
5         1
Name: count, dtype: int64

## 2.8,2.9 Order and Product Cardinality

The order-level analysis shows that the dataset contains **65,752 distinct orders** with between **1 and 5 Order Items per order**.

The average number of Order Items per order is approximately **2.75**, while the median is **3**.

The distribution is:

| Order Items | Number of Orders |
|---:|---:|
| 1 | 19,850 |
| 2 | 11,447 |
| 3 | 11,398 |
| 4 | 11,704 |
| 5 | 11,353 |

The analysis also shows that orders contain between **1 and 5 distinct products**.

Furthermore, the maximum number of distinct `Product Card Id` values associated with a single `Order Item Id` is **1**, confirming that each Order Item is associated with exactly one product.

These findings support the following preliminary relationship structure:

**Customer → Order → Order Item → Product**

with one-to-many relationships from Customer to Order, Order to Order Item, and Product to Order Item.

The difference between the average Order Items per Order and the average distinct Products per Order indicates that some orders may contain multiple Order Item records associated with the same product. This relationship will be investigated further before finalizing the analytical model.

# *Understanding the Relationship Between Customer, Order, Order Item, and Product*

In [22]:
# Select one order as an example
example_order = df["Order Id"].iloc[0]

example = (
    df[df["Order Id"] == example_order]
    [["Order Id",
      "Order Item Id",
      "Customer Id",
      "order date (DateOrders)",
      "Product Card Id",
      "Product Name",
      "Order Item Quantity",
      "Sales",
      "Order Item Product Price"]]
    .sort_values("Order Item Id")
)

example

,Order Id,Order Item Id,Customer Id,order date (DateOrders),Product Card Id,Product Name,Order Item Quantity,Sales,Order Item Product Price
0,77202,180517,20755,1/31/2018 22:56,1360,Smart watch,1,327.75,327.75


In [23]:
order_item_counts = (
    df.groupby("Order Id")["Order Item Id"]
      .nunique()
)

example_order = order_item_counts.idxmax()

example = (
    df[df["Order Id"] == example_order]
    [["Order Id",
      "Order Item Id",
      "Customer Id",
      "order date (DateOrders)",
      "Product Card Id",
      "Product Name"]]
)

example

,Order Id,Order Item Id,Customer Id,order date (DateOrders),Product Card Id,Product Name
20305,5,13,11318,1/1/2015 1:24,403,Nike Men's CJ Elite 2 TD Football Cleat
20680,5,12,11318,1/1/2015 1:24,957,Diamondback Women's Serene Classic Comfort Bi
20695,5,9,11318,1/1/2015 1:24,957,Diamondback Women's Serene Classic Comfort Bi
58046,5,10,11318,1/1/2015 1:24,365,Perfect Fitness Perfect Rip Deck
114914,5,11,11318,1/1/2015 1:24,1014,O'Brien Men's Neoprene Life Vest


In [24]:
# Find customers who placed more than one order
customer_order_counts = (
    df.groupby("Customer Id")["Order Id"]
      .nunique()
      .sort_values(ascending=False)
)

# Select 4 customers with the highest number of orders
top_4_customers = customer_order_counts.head(4).index

# Display their orders
customer_orders_example = (
    df[df["Customer Id"].isin(top_4_customers)]
    [["Customer Id", "Order Id", "order date (DateOrders)"]]
    .drop_duplicates()
    .sort_values(["Customer Id", "order date (DateOrders)"])
)

customer_orders_example

,Customer Id,Order Id,order date (DateOrders)
6224,5654,50910,1/13/2017 3:41
149044,5654,26712,1/25/2016 22:03
77071,5654,68793,10/1/2017 4:54
141590,5654,44352,10/9/2016 10:07
73198,5654,49073,12/17/2016 8:06
51532,5654,28732,2/24/2016 9:44
30223,5654,4730,3/11/2015 0:46
10795,5654,55623,3/22/2017 22:51
150631,5654,32362,4/17/2016 9:29
101390,5654,33113,4/28/2016 8:36


In [25]:
customer_order_summary = (
    df[df["Customer Id"].isin(top_4_customers)]
    .groupby("Customer Id")
    .agg(
        Number_of_Orders=("Order Id", "nunique"),
        Orders=("Order Id", lambda x: list(x.unique()))
    )
    .reset_index()
)

customer_order_summary

,Customer Id,Number_of_Orders,Orders
0,5654,15,"[50910, 65544, 55623, 34004, 4730, 28732, 1176..."
1,10591,14,"[37111, 7727, 33849, 7348, 29001, 59628, 24347..."
2,12284,14,"[2781, 51236, 35497, 52976, 47059, 50357, 1556..."
3,12431,14,"[51865, 29232, 29109, 27585, 46217, 25614, 458..."


In [26]:
customer_example = top_4_customers[0]

customer_orders_items = (
    df[df["Customer Id"] == customer_example]
    [["Customer Id",
      "Order Id",
      "Order Item Id",
      "Product Card Id",
      "order date (DateOrders)",
      "Product Name"]]
    .sort_values(["Order Id", "Order Item Id"])
)

customer_orders_items

,Customer Id,Order Id,Order Item Id,Product Card Id,order date (DateOrders),Product Name
30223,5654,4730,11858,191,3/11/2015 0:46,Nike Men's Free 5.0+ Running Shoe
30236,5654,4730,11859,365,3/11/2015 0:46,Perfect Fitness Perfect Rip Deck
61162,5654,11763,29419,1073,6/21/2015 16:45,Pelican Sunstream 100 Kayak
150152,5654,26712,66894,365,1/25/2016 22:03,Perfect Fitness Perfect Rip Deck
149095,5654,26712,66895,403,1/25/2016 22:03,Nike Men's CJ Elite 2 TD Football Cleat
149171,5654,26712,66896,957,1/25/2016 22:03,Diamondback Women's Serene Classic Comfort Bi
149044,5654,26712,66897,365,1/25/2016 22:03,Perfect Fitness Perfect Rip Deck
149161,5654,26712,66898,957,1/25/2016 22:03,Diamondback Women's Serene Classic Comfort Bi
142113,5654,28732,71923,403,2/24/2016 9:44,Nike Men's CJ Elite 2 TD Football Cleat
142133,5654,28732,71924,818,2/24/2016 9:44,Titleist Pro V1x Golf Balls


## 🔎 Key Finding: Customer → Order → Order Item Relationship

During the data understanding phase, we investigated the relationship between `Customer Id`,
`Order Id`, and `Order Item Id` using real records from the dataset.

A key example is **Customer Id = 5654**, who appears across multiple years and has placed
multiple orders.

### Example from the dataset

| Customer Id | Order Id | Order Item Id | Order Date | Product |
|---:|---:|---:|---|---|
| 5654 | 4730 | 11858 | 2015-03-11 00:46 | Nike Men's Free 5.0+ Running Shoe |
| 5654 | 4730 | 11859 | 2015-03-11 00:46 | Perfect Fitness Perfect Rip Deck |
| 5654 | 26712 | 66894 | 2016-01-25 22:03 | Perfect Fitness Perfect Rip Deck |
| 5654 | 26712 | 66895 | 2016-01-25 22:03 | Nike Men's CJ Elite 2 TD Football Cleat |
| 5654 | 26712 | 66896 | 2016-01-25 22:03 | Diamondback Women's Serene Classic Comfort Bi |
| 5654 | 28732 | 71923 | 2016-02-24 09:44 | Nike Men's CJ Elite 2 TD Football Cleat |
| 5654 | 28732 | 71924 | 2016-02-24 09:44 | Titleist Pro V1x Golf Balls |

### What does this demonstrate?

The example confirms three different levels of the transaction structure:

**1. One Customer can place multiple Orders**

Customer `5654` placed different orders over time:

```text
Customer 5654
│
├── Order 4730
├── Order 26712
├── Order 28732
├── Order 32362
└── ...

Order 26712
│
├── Order Item 66894
├── Order Item 66895
├── Order Item 66896
One Customer → Many Orders → Many Order Items → Products

### *2.10 Missing Values*

In [27]:
missing_summary = (
    df.isna()
      .sum()
      .to_frame("Missing Values")
)

missing_summary["Missing %"] = (
    missing_summary["Missing Values"]
    / len(df)
    * 100
)

missing_summary = (
    missing_summary
    .sort_values("Missing Values", ascending=False)
)

missing_summary[missing_summary["Missing Values"] > 0]

,Missing Values,Missing %
Product Description,180519,100.000000
Order Zipcode,155679,86.239676
Customer Lname,8,0.004432
Customer Zipcode,3,0.001662


### *2.11 Duplicate Records*

In [28]:
df.duplicated().sum()

np.int64(0)

In [29]:
df[["Order Id", "Order Item Id"]].duplicated().sum()

np.int64(0)

In [30]:
df[["Order Id", "Order Item Id"]].duplicated().sum()

np.int64(0)

### *2.12 Categorical Variables*

In [31]:
categorical_columns = df.select_dtypes(
    include="object"
).columns.tolist()

categorical_columns

['Type',
 'Delivery Status',
 'Category Name',
 'Customer City',
 'Customer Country',
 'Customer Email',
 'Customer Fname',
 'Customer Lname',
 'Customer Password',
 'Customer Segment',
 'Customer State',
 'Customer Street',
 'Department Name',
 'Market',
 'Order City',
 'Order Country',
 'order date (DateOrders)',
 'Order Region',
 'Order State',
 'Order Status',
 'Product Image',
 'Product Name',
 'shipping date (DateOrders)',
 'Shipping Mode']

In [32]:
categorical_summary = pd.DataFrame({
    "Feature": categorical_columns,
    "Unique Values": [
        df[col].nunique(dropna=False)
        for col in categorical_columns
    ]
}).sort_values(
    "Unique Values",
    ascending=False
)

categorical_summary

,Feature,Unique Values
16,order date (DateOrders),65752
22,shipping date (DateOrders),63701
11,Customer Street,7458
14,Order City,3597
7,Customer Lname,1110
18,Order State,1089
6,Customer Fname,782
3,Customer City,563
15,Order Country,164
20,Product Image,118


In [33]:
df["Order Status"].value_counts(dropna=False)

Order Status
COMPLETE           59491
PENDING_PAYMENT    39832
PROCESSING         21902
PENDING            20227
CLOSED             19616
ON_HOLD             9804
SUSPECTED_FRAUD     4062
CANCELED            3692
PAYMENT_REVIEW      1893
Name: count, dtype: int64

In [34]:
df["Department Name"].value_counts(dropna=False)

Department Name
Fan Shop              66861
Apparel               48998
Golf                  33220
Footwear              14525
Outdoors               9686
Fitness                2479
Discs Shop             2026
Technology             1465
Pet Shop                492
Book Shop               405
Health and Beauty       362
Name: count, dtype: int64

In [35]:
df["Category Name"].value_counts(dropna=False)

Category Name
Cleats                  24551
Men's Footwear          22246
Women's Apparel         21035
Indoor/Outdoor Games    19298
Fishing                 17325
Water Sports            15540
Camping & Hiking        13729
Cardio Equipment        12487
Shop By Sport           10984
Electronics              3156
Accessories              1780
Golf Balls               1475
Girls' Apparel           1201
Golf Gloves              1070
Trade-In                  974
Video Games               838
Children's Clothing       652
Women's Clothing          650
Baseball & Softball       632
Hockey                    614
Cameras                   592
Toys                      529
Golf Shoes                524
Pet Supplies              492
Crafts                    484
Garden                    484
DVDs                      483
Computers                 442
Golf Apparel              441
Hunting & Shooting        440
Music                     434
Consumer Electronics      431
Boxing & MMA              

In [36]:
df["Shipping Mode"].value_counts(dropna=False)

Shipping Mode
Standard Class    107752
Second Class       35216
First Class        27814
Same Day            9737
Name: count, dtype: int64

In [37]:
df["Customer Segment"].value_counts(dropna=False)

Customer Segment
Consumer       93504
Corporate      54789
Home Office    32226
Name: count, dtype: int64

In [38]:
df["Market"].value_counts(dropna=False)

Market
LATAM           51594
Europe          50252
Pacific Asia    41260
USCA            25799
Africa          11614
Name: count, dtype: int64

In [39]:
df["Delivery Status"].value_counts(dropna=False)

Delivery Status
Late delivery        98977
Advance shipping     41592
Shipping on time     32196
Shipping canceled     7754
Name: count, dtype: int64

In [40]:
df["Order Region"].value_counts(dropna=False)

Order Region
Central America    28341
Western Europe     27109
South America      14935
Oceania            10148
Northern Europe     9792
Southeast Asia      9539
Southern Europe     9431
Caribbean           8318
West of USA         7993
South Asia          7731
Eastern Asia        7280
East of USA         6915
West Asia           6009
US Center           5887
South of  USA       4045
Eastern Europe      3920
West Africa         3696
North Africa        3232
East Africa         1852
Central Africa      1677
Southern Africa     1157
Canada               959
Central Asia         553
Name: count, dtype: int64

### *2.13 Numerical Variables*

In [42]:
import numpy as np
numeric_columns = df.select_dtypes(
    include=np.number
).columns.tolist()

numeric_columns

['Days for shipping (real)',
 'Days for shipment (scheduled)',
 'Benefit per order',
 'Sales per customer',
 'Late_delivery_risk',
 'Category Id',
 'Customer Id',
 'Customer Zipcode',
 'Department Id',
 'Latitude',
 'Longitude',
 'Order Customer Id',
 'Order Id',
 'Order Item Cardprod Id',
 'Order Item Discount',
 'Order Item Discount Rate',
 'Order Item Id',
 'Order Item Product Price',
 'Order Item Profit Ratio',
 'Order Item Quantity',
 'Sales',
 'Order Item Total',
 'Order Profit Per Order',
 'Order Zipcode',
 'Product Card Id',
 'Product Category Id',
 'Product Description',
 'Product Price',
 'Product Status']

In [43]:
df[numeric_columns].describe()

,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Late_delivery_risk,Category Id,Customer Id,Customer Zipcode,Department Id,Latitude,Longitude,Order Customer Id,Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Zipcode,Product Card Id,Product Category Id,Product Description,Product Price,Product Status
count,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180516.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,180519.000000,24840.000000,180519.000000,180519.000000,0.0,180519.000000,180519.0
mean,3.497654,2.931847,21.974989,183.107609,0.548291,31.851451,6691.379495,35921.126914,5.443460,29.719955,-84.915675,6691.379495,36221.894903,692.509764,20.664741,0.101668,90260.000000,141.232550,0.120647,2.127638,203.772096,183.107609,21.974989,55426.132327,692.509764,31.851451,NaN,141.232550,0.0
std,1.623722,1.374449,104.433526,120.043670,0.497664,15.640064,4162.918106,37542.461122,1.629246,9.813646,21.433241,4162.918106,21045.379569,336.446807,21.800901,0.070415,52111.490959,139.732492,0.466796,1.453451,132.273077,120.043670,104.433526,31919.279101,336.446807,15.640064,NaN,139.732492,0.0
min,0.000000,0.000000,-4274.979980,7.490000,0.000000,2.000000,1.000000,603.000000,2.000000,-33.937553,-158.025986,1.000000,1.000000,19.000000,0.000000,0.000000,1.000000,9.990000,-2.750000,1.000000,9.990000,7.490000,-4274.979980,1040.000000,19.000000,2.000000,NaN,9.990000,0.0
25%,2.000000,2.000000,7.000000,104.379997,0.000000,18.000000,3258.500000,725.000000,4.000000,18.265432,-98.446312,3258.500000,18057.000000,403.000000,5.400000,0.040000,45130.500000,50.000000,0.080000,1.000000,119.980003,104.379997,7.000000,23464.000000,403.000000,18.000000,NaN,50.000000,0.0
50%,3.000000,4.000000,31.520000,163.990005,1.000000,29.000000,6457.000000,19380.000000,5.000000,33.144863,-76.847908,6457.000000,36140.000000,627.000000,14.000000,0.100000,90260.000000,59.990002,0.270000,1.000000,199.919998,163.990005,31.520000,59405.000000,627.000000,29.000000,NaN,59.990002,0.0
75%,5.000000,4.000000,64.800003,247.399994,1.000000,45.000000,9779.000000,78207.000000,7.000000,39.279617,-66.370583,9779.000000,54144.000000,1004.000000,29.990000,0.160000,135389.500000,199.990005,0.360000,3.000000,299.950012,247.399994,64.800003,90008.000000,1004.000000,45.000000,NaN,199.990005,0.0
max,6.000000,4.000000,911.799988,1939.989990,1.000000,76.000000,20757.000000,99205.000000,12.000000,48.781933,115.263077,20757.000000,77204.000000,1363.000000,500.000000,0.250000,180519.000000,1999.989990,0.500000,5.000000,1999.989990,1939.989990,911.799988,99301.000000,1363.000000,76.000000,NaN,1999.989990,0.0


### *2.14 Date & Time Understanding*

In [44]:
df["order date (DateOrders)"] = pd.to_datetime(
    df["order date (DateOrders)"],
    errors="coerce"
)

In [45]:
df["order date (DateOrders)"].min()

Timestamp('2015-01-01 00:00:00')

In [46]:
df["order date (DateOrders)"].max()

Timestamp('2018-01-31 23:38:00')

In [47]:
df["order date (DateOrders)"].dt.year.value_counts().sort_index()

order date (DateOrders)
2015    62650
2016    62550
2017    53196
2018     2123
Name: count, dtype: int64

In [48]:
df["order date (DateOrders)"].dt.month.value_counts().sort_index()

order date (DateOrders)
1     17979
2     14529
3     15919
4     15435
5     15976
6     15139
7     15922
8     15912
9     15489
10    12955
11    12500
12    12764
Name: count, dtype: int64

# 2.16 Phase 2 — Summary

The Data Understanding phase established the structural characteristics, transactional grain, identifier behavior, and initial analytical feasibility of the DataCo Supply Chain dataset.

### Key Findings

- The dataset contains **180,519 transactional records**.
- There are **65,752 distinct orders**.
- There are **20,652 distinct customers**.
- There are **118 distinct products**.
- `Order Item Id` uniquely identifies the transactional row.
- One customer can place multiple orders.
- One order can contain multiple order items.
- Each order item is associated with one product.
- The confirmed transactional hierarchy is:

**Customer → Order → Order Item → Product**

### Data Understanding Priorities

The analysis also identified the need to validate:

- Identifier integrity
- Missing-value patterns
- Duplicate records
- Categorical consistency
- Numerical value ranges
- Date and time consistency
- Availability and reliability of variables required for the business KPIs

### Business Relevance

The findings from this phase are critical for the next stages because KPI calculations must respect the dataset grain.

For example:

- Order volume must be calculated using distinct `Order Id`.
- Customer volume must be calculated using distinct `Customer Id`.
- Transaction volume must be calculated using `Order Item Id`.
- Product analysis must use the appropriate product identifier.

No business conclusions or data corrections will be finalized until the data-quality assessment is completed.

### Next Phase

**Phase 3 — Data Quality Assessment**

The next phase will systematically investigate data inconsistencies, missing values, invalid categories, numerical anomalies, date issues, and referential integrity before analytical modeling and KPI calculation.

PHASE 1
Business Problem
      ↓
Business Objectives
      ↓
KPIs
      ↓
PHASE 2
Data Understanding
      ↓
Dataset Structure
      ↓
Data Grain
      ↓
Identifiers
      ↓
Relationships
      ↓
Data Quality Discovery
      ↓
KPI Feasibility
      ↓
PHASE 3
Data Quality Assessment

# Phase 3 — Data Quality Assessment

## Objective

The objective of this phase is to assess whether the dataset is sufficiently reliable and consistent for business analysis and KPI calculation.

The assessment focuses on identifying data-quality issues that could affect the accuracy of our three primary business objectives:

1. **Late Delivery Risk**
2. **Fraud & Suspected Fraud Risk**
3. **Profitability Erosion**

The assessment will cover the following dimensions:

- Completeness
- Uniqueness
- Validity
- Consistency
- Integrity
- Temporal consistency
- Business-rule consistency

Importantly, identifying a data-quality issue does not automatically mean that the data should be deleted or corrected.

Each issue will first be investigated from a business perspective before deciding whether it should be:

- Corrected
- Transformed
- Excluded from a specific analysis
- Retained as a legitimate business case
- Or documented as a data limitation

### *3.2 Completeness — Missing Values*

In [49]:
missing_summary = (
    df.isna()
      .sum()
      .to_frame("Missing Values")
)

missing_summary["Missing %"] = (
    missing_summary["Missing Values"] / len(df) * 100
)

missing_summary = (
    missing_summary
    .sort_values("Missing Values", ascending=False)
)

missing_summary[missing_summary["Missing Values"] > 0]

,Missing Values,Missing %
Product Description,180519,100.000000
Order Zipcode,155679,86.239676
Customer Lname,8,0.004432
Customer Zipcode,3,0.001662


### Missing Value Diagnosis & Handling Strategy

During our initial audit, four columns were identified with missing attributes:

1. **`Product Description` (100.00% Missing):** Entirely empty. Dropping to reduce structural clutter.
2. **`Order Zipcode` (86.24% Missing):** Extremely high sparsity. Dropping as geographic routing is already captured via `Order City`, `Order State`, and `Order Country`.
3. **`Customer Lname` (0.004% Missing):** Negligible missing count (8 records). Filling with `'Unknown'` to preserve row integrity.
4. **`Customer Zipcode` (0.002% Missing):** Imputing conditionally based on matching `Customer City` and `Customer Country` mode/first values.

#### Strategic Rationale:
Dropping high-sparsity columns prevents unnecessary memory load and model noise, while smart imputation for customer attributes preserves item-level transactions for accurate KPI calculation.

### *3.3 Missing Values by Business Importance*

In [50]:
late_delivery_columns = [
    "Days for shipping (real)",
    "Days for shipment (scheduled)",
    "Delivery Status"
]

df[late_delivery_columns].isna().sum()

Days for shipping (real)         0
Days for shipment (scheduled)    0
Delivery Status                  0
dtype: int64

In [51]:
profitability_columns = [
    "Sales",
    "Order Item Profit Ratio",
    "Order Item Discount",
    "Shipping Mode"
]

df[profitability_columns].isna().sum()

Sales                      0
Order Item Profit Ratio    0
Order Item Discount        0
Shipping Mode              0
dtype: int64

In [52]:
fraud_columns = [
    "Order Status"
]

df[fraud_columns].isna().sum()

Order Status    0
dtype: int64

### *3.4 Duplicate Records*

In [53]:
df.duplicated().sum()

np.int64(0)

*No duplicate rows does not necessarily mean no duplicate business transactions.*

### *3.5 Order Integrity*

In [54]:
order_customer_check = (
    df.groupby("Order Id")["Customer Id"]
      .nunique()
)

order_customer_check.value_counts().sort_index()

Customer Id
1    65752
Name: count, dtype: int64

### *3.6 Order Date Consistency*

In [55]:
order_date_check = (
    df.groupby("Order Id")["order date (DateOrders)"]
      .nunique()
)

order_date_check.value_counts().sort_index()

order date (DateOrders)
1    65752
Name: count, dtype: int64

### *3.7 Product Integrity*

In [56]:
product_name_check = (
    df.groupby("Product Card Id")["Product Name"]
      .nunique()
)

product_name_check.value_counts().sort_index()

Product Name
1    118
Name: count, dtype: int64

In [57]:
df.groupby("Product Card Id")["Category Name"].nunique()

Product Card Id
19      1
24      1
35      1
37      1
44      1
       ..
1359    1
1360    1
1361    1
1362    1
1363    1
Name: Category Name, Length: 118, dtype: int64

In [58]:
df.groupby("Product Card Id")["Department Name"].nunique()

Product Card Id
19      1
24      1
35      1
37      1
44      1
       ..
1359    1
1360    1
1361    1
1362    1
1363    1
Name: Department Name, Length: 118, dtype: int64

### *3.8 Customer Integrity*

In [59]:
customer_name_check = (
    df.groupby("Customer Id")["Customer Fname"].nunique()
)

customer_name_check[customer_name_check > 1].head()

Series([], Name: Customer Fname, dtype: int64)

In [60]:
df.groupby("Customer Id")["Customer Lname"].nunique()

Customer Id
1        1
2        1
3        1
4        1
5        1
        ..
20753    1
20754    1
20755    1
20756    1
20757    1
Name: Customer Lname, Length: 20652, dtype: int64

In [61]:
df.groupby("Customer Id")["Customer Country"].nunique()

Customer Id
1        1
2        1
3        1
4        1
5        1
        ..
20753    1
20754    1
20755    1
20756    1
20757    1
Name: Customer Country, Length: 20652, dtype: int64

In [62]:
df.groupby("Customer Id")["Customer City"].nunique()

Customer Id
1        1
2        1
3        1
4        1
5        1
        ..
20753    1
20754    1
20755    1
20756    1
20757    1
Name: Customer City, Length: 20652, dtype: int64

## 3.1 Data Quality Issue Register

| Quality Dimension | Check | Result | Business Impact | Action |
|---|---|---|---|---|
| Completeness | Missing values | Investigated | Depends on feature | Assess by business context |
| Uniqueness | Duplicate rows | To validate | Potential transaction duplication | Investigate |
| Identifier Integrity | Order Item uniqueness | To validate | Transaction grain | Validate |
| Referential Integrity | Order → Customer | To validate | Customer/order analysis | Investigate |
| Referential Integrity | Order Item → Product | To validate | Product analysis | Investigate |
| Categorical Validity | Status fields | To validate | KPI classification | Standardize if required |
| Numerical Validity | Quantity | To validate | Sales calculations | Investigate invalid values |
| Numerical Validity | Sales | To validate | Revenue KPIs | Investigate negative/extreme values |
| Delivery Consistency | Real vs Scheduled Days | To validate | Late delivery KPI | Validate business rule |
| Fraud Classification | Suspected Fraud | To validate | Fraud KPI | Validate at order level |

## 3.18 Data Quality Assessment Conclusion

The dataset has been evaluated across multiple data-quality dimensions, including completeness, uniqueness, identifier integrity, referential integrity, categorical consistency, numerical validity, temporal consistency, and business-rule consistency.

The purpose of this assessment is not simply to identify errors, but to determine whether each irregularity represents:

- A genuine data-quality issue
- A legitimate business scenario
- An expected missing value
- An analytical limitation
- Or a value requiring transformation before analysis

Particular attention is given to the dataset grain:

**Customer → Order → Order Item → Product**

This ensures that subsequent KPI calculations are performed at the correct analytical level.

The final decision regarding data preparation will be based on the business impact of each identified issue rather than applying generic cleaning rules.

### *Phase 4 — Data Preparation* <br>

- ├── 4.1 Preparation Strategy
- ├── 4.2 Create Working Copy
- ├── 4.3 Data Type Standardization
- ├── 4.4 Date & Time Transformation
- ├── 4.5 Missing Value Treatment
- ├── 4.6 Categorical Standardization
- ├── 4.7 Numerical Data Treatment
- ├── 4.8 Identifier Preparation
- ├── 4.9 Business Rule Features
- ├── 4.10 KPI-Ready Features
- ├── 4.11 Validation After Transformation
- └── 4.12 Phase 4 Summary

### *4.1 Data Preparation Strategy*

# Phase 4 — Data Preparation & Transformation

## Objective

The objective of this phase is to transform the validated raw dataset into a reliable, consistent, and analysis-ready dataset.

The preparation process is driven by the business objectives established in Phase 1:

1. Late Delivery Risk
2. Fraud & Suspected Fraud Risk
3. Profitability Erosion

Data preparation will focus on:

- Standardizing data types
- Handling missing values according to business context
- Standardizing categorical variables
- Validating and transforming numerical variables
- Preserving the correct transactional grain
- Creating business-oriented analytical features
- Preparing the dataset for exploratory analysis and KPI calculation

### Important Principle

The raw dataset will remain unchanged.

All transformations will be performed on a working copy so that the original source remains available for auditing and reproducibility.

No record will be removed unless a clear business or analytical justification has been established.

### *4.2 Create a Working Copy*

In [63]:
df_raw = df.copy()
df_clean = df.copy()
df_raw.shape, df_clean.shape

((180519, 53), (180519, 53))

### *4.3 Data Type Standardization*

In [64]:
df_clean.dtypes

Type                                     object
Days for shipping (real)                  int64
Days for shipment (scheduled)             int64
Benefit per order                       float64
Sales per customer                      float64
Delivery Status                          object
Late_delivery_risk                        int64
Category Id                               int64
Category Name                            object
Customer City                            object
Customer Country                         object
Customer Email                           object
Customer Fname                           object
Customer Id                               int64
Customer Lname                           object
Customer Password                        object
Customer Segment                         object
Customer State                           object
Customer Street                          object
Customer Zipcode                        float64
Department Id                           

In [103]:
df_clean["order date (DateOrders)"] = pd.to_datetime(
    df_clean["order date (DateOrders)"],
    errors="coerce"
)

In [104]:
df_clean["order date (DateOrders)"].dtype

dtype('<M8[ns]')

### *4.4 Date & Time Features*

In [105]:
df_clean["order_year"] = (
    df_clean["order date (DateOrders)"].dt.year
)

df_clean["order_month"] = (
    df_clean["order date (DateOrders)"].dt.month
)

df_clean["order_month_name"] = (
    df_clean["order date (DateOrders)"].dt.month_name()
)

df_clean["order_day"] = (
    df_clean["order date (DateOrders)"].dt.day
)

df_clean["order_day_of_week"] = (
    df_clean["order date (DateOrders)"].dt.dayofweek
)

df_clean["order_day_name"] = (
    df_clean["order date (DateOrders)"].dt.day_name()
)
df_clean["order_quarter"] = (
    df_clean["order date (DateOrders)"].dt.quarter
)

In [109]:
df_clean.to_csv("D:\\DataCo Supply Chain.csv",index=False)

### *4.5 Missing Value Treatment*

In [94]:
critical_columns = [
    "Order Id",
    "Order Item Id",
    "Customer Id",
    "Product Card Id",
    "order date (DateOrders)"
]

df_clean[critical_columns].isna().sum()

Order Id                   0
Order Item Id              0
Customer Id                0
Product Card Id            0
order date (DateOrders)    0
dtype: int64

In [95]:
# Create a copy of the raw data to maintain lineage
df_clean = df.copy()

print(f"Initial Shape: {df_clean.shape}")

# 1. Drop irrecoverable and high-sparsity columns
cols_to_drop = ["Product Description", "Order Zipcode"]
df_clean.drop(columns=cols_to_drop, inplace=True, errors="ignore")

# 2. Fill missing Customer Last Name with 'Unknown'
df_clean["Customer Lname"].fillna("Unknown", inplace=True)

# 3. Smart Imputation for Customer Zipcode based on Country and City
# First, fill using the mode zipcode for the same Customer City & Country
zip_lookup = (
    df_clean.groupby(["Customer Country", "Customer City"])["Customer Zipcode"]
    .transform(lambda x: x.mode()[0] if not x.mode().empty else np.nan)
)

df_clean["Customer Zipcode"] = df_clean["Customer Zipcode"].fillna(zip_lookup)

# If any Zipcode remains null, fill with 'Unknown' or overall Country fallback
df_clean["Customer Zipcode"].fillna("00000", inplace=True)

# 4. Verify that missing values are completely resolved
missing_after = df_clean[
    ["Customer Lname", "Customer Zipcode"]
].isnull().sum()

print("\n--- Missing Values Audit After Handling ---")
print(missing_after)
print(f"\nFinal Cleaned Shape: {df_clean.shape}")

Initial Shape: (180519, 53)


C:\Users\manti\AppData\Local\Temp\ipykernel_2512\2023435873.py:11: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_clean["Customer Lname"].fillna("Unknown", inplace=True)
C:\Users\manti\AppData\Local\Temp\ipykernel_2512\2023435873.py:23: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy


--- Missing Values Audit After Handling ---
Customer Lname      0
Customer Zipcode    0
dtype: int64

Final Cleaned Shape: (180519, 51)


### *4.6 Categorical Standardization*

In [96]:
categorical_columns = df_clean.select_dtypes(
    include="object"
).columns

for col in categorical_columns:
    df_clean[col] = df_clean[col].str.strip()

In [97]:
df_clean["Order Status"].unique()

array(['COMPLETE', 'PENDING', 'CLOSED', 'PENDING_PAYMENT', 'CANCELED',
       'PROCESSING', 'SUSPECTED_FRAUD', 'ON_HOLD', 'PAYMENT_REVIEW'],
      dtype=object)

In [98]:
df_clean["Shipping Mode"] = (
    df_clean["Shipping Mode"]
    .str.strip()
    .str.title()
)

### *4.7 Create Business Features*

In [99]:
df_clean["shipping_delay_days"] = (
    df_clean["Days for shipping (real)"]
    - df_clean["Days for shipment (scheduled)"]
)

In [100]:
df_clean["is_late_delivery"] = (
    df_clean["shipping_delay_days"] > 0
)

# 4.16 Phase 4 — Data Preparation Summary

The Data Preparation phase transformed the validated dataset into an analysis-ready structure while preserving the original raw data.

### Main transformations performed

- Created a separate working dataset while preserving the raw source.
- Standardized relevant data types.
- Converted order timestamps into proper datetime format.
- Created analytical date attributes such as year, month, quarter, and day of week.
- Standardized categorical text formatting where required.
- Converted relevant numerical fields into appropriate numeric types.
- Preserved identifiers as identifiers rather than treating them as numerical measures.
- Investigated missing values according to their business meaning rather than applying blanket imputation.
- Preserved legitimate business outliers unless evidence indicated that they were data errors.

### Business-Oriented Features

Several analytical features were prepared to support the project's business objectives:

- `shipping_delay_days`
- `is_late_delivery`

These features will support the analysis of:

- Late Delivery Risk

### Analytical Aggregations

Additional analytical datasets were prepared at different business levels:

- Order level
- Customer level
- Product level

This was necessary because the original transactional grain is at the **Order Item** level, while several business questions require analysis at the order, customer, or product level.

### Validation

After transformation, the dataset was revalidated to ensure:

- No unintended row loss
- Identifier integrity was preserved
- Duplicate transactions were not introduced
- Required identifiers remained available
- New features contained logically valid values

The resulting dataset is now prepared for the next analytical stage.

### Next Phase

**Phase 5 — Exploratory Data Analysis (EDA) & Business KPI Analysis**

The next phase will move from data preparation to business discovery, beginning with overall sales, order, customer, delivery, fraud, and profitability performance.

Star Schema ERD Visual Architecture
                  +-----------------------------------+
                  |           Dim_Customer            |
                  +-----------------------------------+
                  | PK  Customer_SK                   |
                  |     Customer Id (Business Key)    |
                  |     Customer Fname, Customer Lname|
                  |     Customer Email, Password      |
                  |     Customer Segment              |
                  |     Customer Street, City, State  |
                  |     Customer Country, Zipcode     |
                  |     Latitude, Longitude           |
                  +-----------------------------------+
                                    | 1
                                    |
                                    | *
+-----------------------+   +---------------------------------------+   +-----------------------------------+
|       Dim_Date        |   |           Fact_Order_Items            |   |            Dim_Product            |
+-----------------------+   +---------------------------------------+   +-----------------------------------+
| PK  Date_Key          | 1 | PK  Order_Item_SK (Surrogate Key)     | * | PK  Product_SK                    |
|     Date (YYYY-MM-DD) |---| FK  Order_Date_Key                    |---|     Product Card Id (Business Key)|
|     Year, Quarter     | * | FK  Shipping_Date_Key                 | 1 |     Product Name, Product Price   |
|     Month, Month Name |   | FK  Customer_SK                       |   |     Product Status, Description   |
|     Day, Day of Week  |   | FK  Product_SK                        |   |     Product Image                 |
|     Is Weekend        |   | FK  Location_SK                       |   |     Category Id, Category Name    |
+-----------------------+   | FK  Shipping_SK                       |   |     Department Id, Department Name|
                            | FK  Order_Status_SK                   |   +-----------------------------------+
                            |     Order Id (Degenerate Key)         |
                            |     Type (Payment Method)             |
                            |     Days for shipping (real)          |
                            |     Days for shipment (scheduled)     |
                            |     Late_delivery_risk                |
                            |     Order Item Quantity               |
                            |     Order Item Product Price          |
                            |     Order Item Discount               |
                            |     Order Item Discount Rate          |
                            |     Order Item Total (Sales)          |
                            |     Order Item Profit Ratio           |
                            |     Order Profit Per Order            |
                            |     Sales per customer                |
                            +---------------------------------------+
                                 | *                      * |
                                 |                          |
                                 | 1                      1 |
                  +------------------------------+   +-----------------------------------+
                  |         Dim_Shipping         |   |           Dim_Location            |
                  +------------------------------+   +-----------------------------------+
                  | PK  Shipping_SK              |   | PK  Location_SK                   |
                  |     Shipping Mode            |   |     Order City, Order State       |
                  |     Delivery Status          |   |     Order Country, Order Region   |
                  +------------------------------+   |     Order Zipcode, Market         |
                                                     +-----------------------------------+

In [78]:
df_clean.to_csv('D:\\DataCo Smart Supply Chain Risk.csv',index=False)

In [101]:
df_clean

,Type,Days for shipping (real),Days for shipment (scheduled),Benefit per order,Sales per customer,Delivery Status,Late_delivery_risk,Category Id,Category Name,Customer City,Customer Country,Customer Email,Customer Fname,Customer Id,Customer Lname,Customer Password,Customer Segment,Customer State,Customer Street,Customer Zipcode,Department Id,Department Name,Latitude,Longitude,Market,Order City,Order Country,Order Customer Id,order date (DateOrders),Order Id,Order Item Cardprod Id,Order Item Discount,Order Item Discount Rate,Order Item Id,Order Item Product Price,Order Item Profit Ratio,Order Item Quantity,Sales,Order Item Total,Order Profit Per Order,Order Region,Order State,Order Status,Product Card Id,Product Category Id,Product Image,Product Name,Product Price,Product Status,shipping date (DateOrders),Shipping Mode,shipping_delay_days,is_late_delivery
0,DEBIT,3,4,91.250000,314.640015,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Cally,20755,Holloway,XXXXXXXXX,Consumer,PR,5365 Noble Nectar Island,NaN,2,Fitness,18.251453,-66.037056,Pacific Asia,Bekasi,Indonesia,20755,2018-01-31 22:56:00,77202,1360,13.110000,0.04,180517,327.750000,0.29,1,327.750000,314.640015,91.250000,Southeast Asia,Java Occidental,COMPLETE,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.750000,0,2/3/2018 22:56,Standard Class,-1,False
1,TRANSFER,5,4,-249.089996,311.359985,Late delivery,1,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Irene,19492,Luna,XXXXXXXXX,Consumer,PR,2679 Rustic Loop,NaN,2,Fitness,18.279451,-66.037064,Pacific Asia,Bikaner,India,19492,2018-01-13 12:27:00,75939,1360,16.389999,0.05,179254,327.750000,-0.80,1,327.750000,311.359985,-249.089996,South Asia,Rajastán,PENDING,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.750000,0,1/18/2018 12:27,Standard Class,1,True
2,CASH,4,4,-247.779999,309.720001,Shipping on time,0,73,Sporting Goods,San Jose,EE. UU.,XXXXXXXXX,Gillian,19491,Maldonado,XXXXXXXXX,Consumer,CA,8510 Round Bear Gate,NaN,2,Fitness,37.292233,-121.881279,Pacific Asia,Bikaner,India,19491,2018-01-13 12:06:00,75938,1360,18.030001,0.06,179253,327.750000,-0.80,1,327.750000,309.720001,-247.779999,South Asia,Rajastán,CLOSED,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.750000,0,1/17/2018 12:06,Standard Class,0,False
3,DEBIT,3,4,22.860001,304.809998,Advance shipping,0,73,Sporting Goods,Los Angeles,EE. UU.,XXXXXXXXX,Tana,19490,Tate,XXXXXXXXX,Home Office,CA,3200 Amber Bend,NaN,2,Fitness,34.125946,-118.291016,Pacific Asia,Townsville,Australia,19490,2018-01-13 11:45:00,75937,1360,22.940001,0.07,179252,327.750000,0.08,1,327.750000,304.809998,22.860001,Oceania,Queensland,COMPLETE,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.750000,0,1/16/2018 11:45,Standard Class,-1,False
4,PAYMENT,2,4,134.210007,298.250000,Advance shipping,0,73,Sporting Goods,Caguas,Puerto Rico,XXXXXXXXX,Orli,19489,Hendricks,XXXXXXXXX,Corporate,PR,8671 Iron Anchor Corners,NaN,2,Fitness,18.253769,-66.037048,Pacific Asia,Townsville,Australia,19489,2018-01-13 11:24:00,75936,1360,29.500000,0.09,179251,327.750000,0.45,1,327.750000,298.250000,134.210007,Oceania,Queensland,PENDING_PAYMENT,1360,73,http://images.acmesports.sports/Smart+watch,Smart watch,327.750000,0,1/15/2018 11:24,Standard Class,-2,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
180514,CASH,4,4,40.000000,399.980011,Shipping on time,0,45,Fishing,Brooklyn,EE. UU.,XXXXXXXXX,Maria,1005,Peterson,XXXXXXXXX,Home Office,NY,1322 Broad Glade,NaN,7,Fan Shop,40.640930,-73.942711,Pacific Asia,Shanghái,China,1005,2016-01-16 03:40:00,26043,1004,0.000000,0.00,65177,399.980011,0.10,1,399.980011,399.980011,40.000000,Eastern Asia,Shanghái,CLOSED,1004,45,http://images.acmesports.sports/Field+%26+Stre...,Field & Stream Sportsman 16 Gun Fire Safe,399.980011,0,1/20/2016 3: